### PR TITLE
feat: Allow extra commands in CodeBuildImageBuilder via build argument

### PR DESCRIPTION
**Pull request by bubenkoff**

### CODE REVIEW

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-ARG BASE_IMAGE="public.ecr.aws/lts/ubuntu:20.04"
+ARG BASE_IMAGE="public.ecr.aws/lts/ubuntu:latest"
```
Comment (kichik) [inline src/providers/docker-images/codebuild/linux-arm64/Dockerfile]: Is upgrading required for this change? We should ideally keep this to a separate PR as it may be a breaking change for some people.

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile [hunk 2]
```diff
@@ -65,3 +68,6 @@ VOLUME /var/lib/docker
 
 # configure runner
 USER runner
+
+ARG EXTRA_COMMANDS="echo 'no extra commands'"
+RUN bash -c "${EXTRA_COMMANDS}"
```
Comment (kichik) [inline src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: Will this work with commands containing quotes? Can we maybe use an asset instead where the user just gives us a shell script to execute?

**Review by kichik — COMMENTED** [review]
Thanks. Definitely something that will make life easier for users. Besides the comment, my only other suggestion would be naming it pre and post hooks or something similar that makes it clear they are called before and after anything else is installed.

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 1 above)
Comment (bubenkoff) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile]: not required, but as more recent LTS is already there, I thought it's good always to have it the latest available LTS
If there's some breaking change, can't they put in a local older version of it?

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 2 above)
Comment (bubenkoff) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: with the quotes, it's tricky indeed, but I don't see any other way to use RUN with some arbitrary command like that
an asset is not ideal due to the intended (see example in docs) use case: ad-hoc operations like docker login

**Comment by bubenkoff** [comment]
@kichik do you mean 
PRE_INSTALL_COMMANDS, POST_INSTALL_COMMAND?
PRE_INSTALL_HOOKS, POST_INSTALL_HOOK?
PRE_BUILD_COMMANDS, POST_BUILD_COMMAND?
PRE_BUILD_HOOKS, POST_BUILD_HOOK?
just realized it should be singular not plural as it's used in a single RUN

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 1 above)
Comment (kichik) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile]: Introducing breaking changes should be documented. Yes, they can input an older version, but I don't want that requirement to be a surprise. That's why it should be a separate PR so it can be discussed and documented.

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 2 above)
Comment (kichik) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: We can have an inline asset for this purpose just like `aws_lambda.Code.fromInline`. I would prefer to not have gotchas like not supporting quotes in such a useful feature.

**Comment by kichik** [comment]
> @kichik do you mean 
> PRE_INSTALL_COMMANDS, POST_INSTALL_COMMAND?
> PRE_INSTALL_HOOKS, POST_INSTALL_HOOK?
> PRE_BUILD_COMMANDS, POST_BUILD_COMMAND?
> PRE_BUILD_HOOKS, POST_BUILD_HOOK?
> just realized it should be singular not plural as it's used in a single RUN
> 

Well, I think we can start with just pre everything and post everything. I don't know that we would need it for every step. But just something in the name that makes it clear it's before and after everything plus an explanation in the comment.

**Comment by bubenkoff** [comment]
> > @kichik do you mean
> > PRE_INSTALL_COMMANDS, POST_INSTALL_COMMAND?
> > PRE_INSTALL_HOOKS, POST_INSTALL_HOOK?
> > PRE_BUILD_COMMANDS, POST_BUILD_COMMAND?
> > PRE_BUILD_HOOKS, POST_BUILD_HOOK?
> > just realized it should be singular not plural as it's used in a single RUN
> 
> Well, I think we can start with just pre everything and post everything. I don't know that we would need it for every step. But just something in the name that makes it clear it's before and after everything plus an explanation in the comment.

I just put the name options to choose from, what do you prefer?

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 2 above)
Comment (bubenkoff) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: how do I implement it then exactly? Do you have any examples?

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 1 above)
Comment (bubenkoff) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile]: no problem, will separate

File: src/providers/docker-images/codebuild/linux-arm64/Dockerfile (hunk 2 above)
Comment (kichik) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: I can add more details later but basically:
```
    new s3_assets.Asset(this, 'Dependencies Definition', {
      path: '.',
      assetHashType: AssetHashType.OUTPUT,
      bundling: {
        local: {
          tryBundle(outputDir: string, _: BundlingOptions): boolean {
            fs.writeFileSync(join(outputDir, 'hook.sh'), 'echo hello world');
            return true;
          },
        },
        // no fallback
        image: DockerImage.fromRegistry('public.ecr.aws/docker/library/busybox:stable'),
        command: ['exit 1'],
      },
    })
```
And then you would have a class like [Code](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_lambda.Code.html) that gives you all the options.

Comment (bubenkoff) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: I'd need more help I'm afraid, inline code + base code class is one thing, but how exactly will it work with getting the value

Comment (kichik) [reply src/providers/docker-images/codebuild/linux-arm64/Dockerfile:73]: I'm not exactly sure what you mean by "value" here. Are you talking about getting the command to run into `Dockerfile`? We can maybe do something similar to the way [certificates are added](https://github.com/CloudSnorkel/cdk-github-runners/blob/206bed8c19b7c555f231d6306cbe0d15c08d4d63/src/providers/image-builders/codebuild.ts#L275-L280). It writes the file to a special folder and then `Dockerfile` looks for files in that folder and [only installs the certificates](https://github.com/CloudSnorkel/cdk-github-runners/blob/206bed8c19b7c555f231d6306cbe0d15c08d4d63/src/providers/docker-images/lambda/linux-arm64/Dockerfile#L10) if it's there. We can also try something else if you have a simpler idea.

If you want, we can do a quick video chat next week to discuss the details and figure out a good solution together. I'm also on the [CDK Slack](https://cdk.dev/) if you prefer chat.

**Comment by bubenkoff** [comment]
@kichik what is the procedure to test this package in part of the CodeBuildimageBuilder?

**Comment by kichik** [comment]
> @kichik what is the procedure to test this package in part of the CodeBuildimageBuilder?

I use the integration test. You can deploy with `npm run integ:default:deploy`. You may want to create a smaller copy of it because initial deployment can take 30 minutes. Just make sure your commit doesn't include those changes or the snapshot change as the build will fail.

**Comment by bubenkoff** [comment]
> > @kichik what is the procedure to test this package in part of the CodeBuildimageBuilder?
> 
> I use the integration test. You can deploy with `npm run integ:default:deploy`. You may want to create a smaller copy of it because initial deployment can take 30 minutes. Just make sure your commit doesn't include those changes or the snapshot change as the build will fail.
getting this one:
```
Error: Cannot find asset at /Users/bubenkoff/projects/cdk-github-runners/lib/lambdas/aws-image-builder-versioner
```

**Comment by kichik** [comment]
> > > @kichik what is the procedure to test this package in part of the CodeBuildimageBuilder?
> > 
> > 
> > I use the integration test. You can deploy with `npm run integ:default:deploy`. You may want to create a smaller copy of it because initial deployment can take 30 minutes. Just make sure your commit doesn't include those changes or the snapshot change as the build will fail.
> > getting this one:
> 
> ```
> Error: Cannot find asset at /Users/bubenkoff/projects/cdk-github-runners/lib/lambdas/aws-image-builder-versioner
> ```

You need to run `npm run compile` or a full `npm run build` every time you change the lambdas. Also see the comments at the top of the integration test file.

**Comment by bubenkoff** [comment]
@kichik sorry I give up on this. For my use case I simply cloned your Dockerfile and modified it the way it's in this PR and it works for me fine. 
Thank you very much for this great project!

**Comment by kichik** [comment]
Thanks for giving it a shot. I think eventually I'm going to create some common interface for both CodeBuild and EC2 Image Builder so you can add components to both. That's feels easier and more confined than files/hooks/commands.

**Comment by bubenkoff** [comment]
Yes I thought the same but as it's quite complex system there I didn't feel
like touching it.
Thanks again
